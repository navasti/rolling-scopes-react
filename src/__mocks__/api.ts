import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  PokemonMove,
  PokemonType,
  Pokemon,
} from 'types';
import { hasError } from 'utils';

export const fetchBaseMock = (mockData: Pokemon | PokemonMove | PokemonType) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = (
  mockData: PokemonDetails | PokemonMoveDetails | PokemonTypeDetails
) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const handleResponseMock = (status: number, mockData: Pokemon) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};

export const handleMappedResponseMock = (status: number, mockData: PokemonDetails) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};
