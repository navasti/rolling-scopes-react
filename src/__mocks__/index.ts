export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
