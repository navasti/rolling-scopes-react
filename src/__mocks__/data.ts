import { ErrorMessages } from 'appConstants';
import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  PokemonData,
  PokemonMove,
  PokemonType,
  Details,
  Pokemon,
  Lengths,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const links = ['Home', 'About', 'Form', 'Invalid'];

export const testingComponentName = 'Test';
export const testingLocation = '/test';

export const testingContent = 'Content';
export const testingTitle = 'Title';

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

//

export const lengths: Lengths = { pokemons: 10, moves: 15, types: 20 };

const samplePokemon = (): PokemonDetails => ({
  base_experience: 62,
  name: 'charmander',
  weight: 85,
  height: 6,
  id: 4,
  abilities: [
    {
      ability: { name: 'blaze', url: 'https://pokeapi.co/api/v2/ability/66/' },
      is_hidden: false,
      slot: 1,
    },
  ],
  sprites: {
    front_default: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
    back_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/4.png',
  },
  stats: [
    {
      base_stat: 52,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
  ],
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'https://pokeapi.co/api/v2/type/10/',
      },
    },
  ],
});

const sampleMove = (): PokemonMoveDetails => ({
  accuracy: 100,
  name: 'tackle',
  priority: 0,
  power: 40,
  id: 33,
  pp: 35,
  learned_by_pokemon: [
    {
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
    },
  ],
  type: {
    name: 'normal',
    url: 'https://pokeapi.co/api/v2/type/1/',
  },
});

const sampleType = (): PokemonTypeDetails => ({
  moves: [{ name: 'gust', url: 'https://pokeapi.co/api/v2/move/16/' }],
  name: 'flying',
  id: 3,
  damage_relations: {
    double_damage_from: [
      { name: 'electric', url: 'https://pokeapi.co/api/v2/type/13/' },
      { name: 'rock', url: 'https://pokeapi.co/api/v2/type/6/' },
    ],
    double_damage_to: [{ name: 'fighting', url: 'https://pokeapi.co/api/v2/type/2/' }],
    half_damage_from: [{ name: 'bug', url: 'https://pokeapi.co/api/v2/type/7/' }],
    half_damage_to: [{ name: 'steel', url: 'https://pokeapi.co/api/v2/type/9/' }],
    no_damage_from: [{ name: 'ground', url: 'https://pokeapi.co/api/v2/type/5/' }],
    no_damage_to: [],
  },
  pokemon: [
    { pokemon: { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' }, slot: 2 },
  ],
});

const prepareBuilder = <T extends Details>(sample: () => T) => {
  const builder = (object = sample()) => ({
    withName: (name: T['name']) => builder({ ...object, name }),
    withId: (id: T['id']) => builder({ ...object, id }),
    valueOf: () => object,
  });
  return builder;
};

const pokemonBuilder = prepareBuilder(samplePokemon);
const moveBuilder = prepareBuilder(sampleMove);
const typeBuilder = prepareBuilder(sampleType);

export const pokemonsMock: Array<PokemonDetails> = [
  pokemonBuilder().withId(1).withName('bulbasaur'),
  pokemonBuilder().withId(25).withName('pikachu'),
  pokemonBuilder(),
].map((builder) => builder.valueOf());

export const movesMock: Array<PokemonMoveDetails> = [
  moveBuilder().withId(55).withName('water-gun'),
  moveBuilder().withId(50).withName('disable'),
  moveBuilder(),
].map((builder) => builder.valueOf());

export const typesMock: Array<PokemonTypeDetails> = [
  typeBuilder().withId(2).withName('fighting'),
  typeBuilder().withId(5).withName('ground'),
  typeBuilder(),
].map((builder) => builder.valueOf());
