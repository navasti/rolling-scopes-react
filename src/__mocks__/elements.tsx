import { createRef, RefObject } from 'react';
import { testingContent, testingTitle } from './data';

export const modalRef: RefObject<HTMLDivElement> = createRef();
export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const ModalTitle = () => <span data-testid="modal-title">{testingTitle}</span>;
export const ModalContent = () => <div data-testid="modal-content">{testingContent}</div>;
