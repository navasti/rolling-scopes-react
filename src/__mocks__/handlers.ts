import { screen } from '@testing-library/react';
import { AvailableTabs } from 'appConstants';

export const expectByText = jest.fn((text: string) => {
  expect(screen.getByText(text)).toBeInTheDocument();
});

export const handleCloseModal = jest.fn();
export const handleCardSelect = jest.fn();
export const handleOpenModal = jest.fn();
export const onTabClick = jest.fn((availableTab: AvailableTabs) => availableTab);
