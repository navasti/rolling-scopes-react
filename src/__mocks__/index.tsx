import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
