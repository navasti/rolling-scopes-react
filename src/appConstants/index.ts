import { FormFields } from 'types';
import * as yup from 'yup';

export enum Limits {
  pokemon = 20,
  type = 10,
  move = 15,
}

export enum ErrorStatuses {
  success = '2',
  client = '4',
  server = '5',
}

export enum AvailableTabs {
  pokemons = 'pokemons',
  moves = 'moves',
  types = 'types',
}

export enum Fields {
  birthday = 'birthday',
  consent = 'consent',
  gender = 'gender',
  avatar = 'avatar',
  shiny = 'shiny',
  name = 'name',
  type = 'type',
}

export enum ErrorMessages {
  name = 'Name must be at least 2 characters long.',
  consent = 'Regulations consent is required.',
  birthday = 'Birthday is required.',
  type = 'Main type is required.',
  gender = 'Gender is required.',
}

export enum Types {
  fire = 'fire',
  grass = 'grass',
  poison = 'poison',
  flying = 'flying',
  ground = 'ground',
  psycho = 'psycho',
  water = 'water',
  ice = 'ice',
}

export const NONE = 'none';
export const MALE = 'male';
export const FEMALE = 'female';
export const INPUT_VALUE_KEY = 'input-value';
export const SUCCESS_MESSAGE = 'Pokemon added successfuly';
export const LINKS = ['home', 'about', 'form', 'invalid', 'details'] as const;
export const SEARCH_BAR_INSTRUCTIONS = [
  'Type and press enter to search for specific pokemon, type or move.',
  'Clear input and press enter to search for all pokemons, types and moves.',
] as const;

export const TABS = Object.values(AvailableTabs);
export const POKEMON_TYPES = Object.values(Types);

export const FORM_VALIDATION_SCHEMA = yup.object().shape({
  type: yup.string().required(ErrorMessages.type),
  name: yup.string().min(2, ErrorMessages.name).required(ErrorMessages.name),
  gender: yup.string().nullable().required(ErrorMessages.gender),
  avatar: yup
    .mixed()
    .test('fileSize', 'The file is too large', (value) =>
      !value?.length ? true : value?.[0]?.size <= 200000
    ),
  birthday: yup.string().required(ErrorMessages.birthday),
  consent: yup.bool().oneOf([true], ErrorMessages.consent),
  shiny: yup.bool(),
});

const BASE = 'https://pokeapi.co/api/v2';

export const API = {
  POKEMON_LIMIT: `?limit=${Limits.pokemon}`,
  TYPE_LIMIT: `?limit=${Limits.type}`,
  MOVE_LIMIT: `?limit=${Limits.move}`,
  POKEMON: `${BASE}/pokemon`,
  TYPE: `${BASE}/type`,
  MOVE: `${BASE}/move`,
  BASE,
};

export const FIELDS_VALIDATION_BY_NAME = {
  TEXT: [Fields.name, Fields.birthday, Fields.type],
  CHECK: [Fields.consent, Fields.gender],
};

export const DEFAULT_VALUES: FormFields = {
  consent: false,
  avatar: null,
  gender: MALE,
  shiny: false,
  birthday: '',
  name: '',
  type: '',
};
