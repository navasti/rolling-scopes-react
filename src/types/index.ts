import { AvailableTabs, FEMALE, Fields, Limits, MALE, TABS } from 'appConstants';
import { UseFormRegisterReturn } from 'react-hook-form';
import { ChangeEvent, KeyboardEvent, RefObject } from 'react';

export type TextFields = Fields.name | Fields.birthday | Fields.type;
export type CheckFields = Fields.consent | Fields.gender;
export type Lengths = PrepareLengths<typeof TABS>;
export type PrepareLengths<Arr extends typeof TABS> = {
  [P in Arr[number] as P]: number;
};

export enum MessageType {
  error = 'error',
  success = 'success',
}

export type Details = PokemonDetails | PokemonMoveDetails | PokemonTypeDetails;

export type CardsProps = {
  types: Array<PokemonTypeDetails>;
  moves: Array<PokemonMoveDetails>;
  pokemons: Array<PokemonDetails>;
  activeTab: AvailableTabs;
  isLoading: boolean;
};
export type LayoutProps = {
  children: JSX.Element;
  componentName: string;
  location: string;
};

export type TabsProps = {
  onClick: (tab: AvailableTabs) => void;
  activeTab: AvailableTabs;
  options: typeof TABS;
  isLoading: boolean;
  lengths: Lengths;
};

export type SearchBarProps = {
  onKeyDown: (event: KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  inputValue: string;
  isLoading: boolean;
  label: string;
};

export type FormFields = {
  gender: typeof MALE | typeof FEMALE;
  avatar: FileList | null;
  consent: boolean;
  birthday: string;
  shiny: boolean;
  name: string;
  type: string;
};

export type FieldReturnType<T extends Fields> = {
  name: string;
  error?: string;
  value?: string;
} & UseFormRegisterReturn<T>;

export type LimitParams = {
  pokemonLimit: Limits.pokemon;
  typeLimit: Limits.type;
  moveLimit: Limits.move;
};

export type PokemonDetails = {
  abilities: Array<Ability>;
  base_experience: number;
  types: Array<Type>;
  stats: Array<Stat>;
  sprites: Sprites;
  height: number;
  weight: number;
  name: string;
  id: number;
};

export type PokemonData = {
  previous: null | string;
  results: Array<Pokemon>;
  next: null | string;
  count: number;
};

export type TypesData = {
  previous: null | string;
  results: Array<PokemonType>;
  next: null | string;
  count: number;
};

export type MovesData = {
  results: Array<PokemonMove>;
  previous: null | string;
  next: null | string;
  count: number;
};

export type DamageRelations = {
  double_damage_from: Array<PokemonType>;
  double_damage_to: Array<PokemonType>;
  half_damage_from: Array<PokemonType>;
  half_damage_to: Array<PokemonType>;
  no_damage_from: Array<PokemonType>;
  no_damage_to: Array<PokemonType>;
};

export type PokemonTypeDetails = {
  pokemon: Array<{ pokemon: Pokemon; slot: number }>;
  damage_relations: DamageRelations;
  moves: Array<PokemonMove>;
  name: string;
  id: number;
};

export type PokemonType = {
  name: string;
  url: string;
};

export type Pokemon = {
  name: string;
  url: string;
};

export type PokemonMove = {
  name: string;
  url: string;
};

export type PokemonMoveDetails = {
  learned_by_pokemon: Array<Pokemon>;
  type: PokemonType;
  accuracy: number;
  priority: number;
  power: number;
  name: string;
  pp: number;
  id: number;
};

export type Location = {
  name: string;
  url: string;
};

export type LocationDetails = {
  areas: Array<Area>;
  region: Region;
  name: string;
  id: number;
};

export type RegionDetails = {
  locations: Array<Location>;
  id: number;
};

export type AreaDetails = {
  pokemon_encounteres: Array<{ pokemon: Pokemon; slot: number }>;
  game_index: number;
  name: string;
  id: number;
};

export type Region = {
  name: string;
  url: string;
};

export type Area = {
  name: string;
  url: string;
};

export type Ability = {
  ability: {
    name: string;
    url: string;
  };
  is_hidden: boolean;
  slot: number;
};

export type Stat = {
  base_stat: number;
  effort: number;
  stat: {
    name: string;
    url: string;
  };
};

export type Type = {
  slot: number;
  type: {
    name: string;
    url: string;
  };
};

export type Sprites = {
  back_default: string;
  front_default: string;
};

export type CommonFieldType = {
  inputRef: RefObject<HTMLInputElement>;
  errorRef: RefObject<HTMLSpanElement>;
};

export type GenderFieldType = {
  femaleInputRef: RefObject<HTMLInputElement>;
  maleInputRef: RefObject<HTMLInputElement>;
  errorRef: RefObject<HTMLSpanElement>;
};

export type SelectFieldType = {
  selectRef: RefObject<HTMLSelectElement>;
  errorRef: RefObject<HTMLSpanElement>;
};

export type CustomPokemon = {
  gender: typeof MALE | typeof FEMALE;
  avatar: File | null;
  birthday: string;
  shiny: boolean;
  name: string;
  type: string;
};

export type ErrorsObject = {
  birthday: null | string;
  consent: null | string;
  gender: null | string;
  name: null | string;
  type: null | string;
};
