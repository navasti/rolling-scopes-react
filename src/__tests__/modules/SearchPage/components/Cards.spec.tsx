import { PokemonDetails, PokemonMoveDetails, PokemonTypeDetails } from 'types';
import { movesMock, pokemonsMock, typesMock } from '__mocks__/data';
import { screen, render } from '@testing-library/react';
import { AvailableTabs } from 'appConstants';

jest.mock('modules', () => {
  const React = require('react');
  return {
    __esModule: true,
    DetailsModal: React.forwardRef((_: unknown, ref: React.RefObject<HTMLDivElement>) => (
      <div ref={ref}></div>
    )),
  };
});

jest.mock('components', () => ({
  __esModule: true,
  Loader: () => <p>loading</p>,
}));

jest.mock('modules/SearchPage/components', () => {
  const { Cards } = jest.requireActual('modules/SearchPage/components/Cards');
  return {
    __esModule: true,
    Cards,
    MoveCard: ({ move }: { move: PokemonMoveDetails }) => (
      <button data-testid="move-card-mock">
        <p>{move.name}</p>
      </button>
    ),
    TypeCard: ({ type }: { type: PokemonTypeDetails }) => (
      <button data-testid="type-card-mock">
        <p>{type.name}</p>
      </button>
    ),
    PokemonCard: ({ pokemon }: { pokemon: PokemonDetails }) => {
      return (
        <button key={pokemon.id} data-testid="pokemon-card-mock">
          <div>
            <img
              // src={pokemon.sprites.front_default}
              data-testid="pokemon-img-mock"
              alt={pokemon.name}
            />
            <div>
              <span>Base:</span>
              <p>
                Name: <span>{pokemon.name}</span>
              </p>
              <p>
                Types: <span>{pokemon.types[0].type.name}</span>
              </p>
            </div>
          </div>
        </button>
      );
    },
  };
});

const customCardsRender = (
  isLoading = false,
  isData = true,
  tab: AvailableTabs = AvailableTabs.pokemons
) => (
  <></>
  // <Cards
  //   pokemons={isData ? pokemonsMock : []}
  //   moves={isData ? movesMock : []}
  //   types={isData ? typesMock : []}
  //   isLoading={isLoading}
  //   activeTab={tab}
  // />
);

describe.skip('Cards', () => {
  it('loader should be rendered and visible if isLoading prop is set to true', () => {
    render(customCardsRender(true));
    expect(screen.getByText('loading')).toBeInTheDocument();
    expect(screen.getByText('loading')).toBeVisible();
  });
  it('when pokemons are set as an active tab but no pokemons were found, display information about it', () => {
    render(customCardsRender(false, false));
    expect(screen.getByText('No pokemons were found')).toBeInTheDocument();
    expect(screen.getByText('No pokemons were found')).toBeVisible();
  });
  it('when moves are set as an active tab but no moves were found, display information about it', () => {
    render(customCardsRender(false, false, AvailableTabs.moves));
    expect(screen.getByText('No moves were found')).toBeInTheDocument();
    expect(screen.getByText('No moves were found')).toBeVisible();
  });
  it('when types are set as an active tab but no types were found, display information about it', () => {
    render(customCardsRender(false, false, AvailableTabs.types));
    expect(screen.getByText('No types were found')).toBeInTheDocument();
    expect(screen.getByText('No types were found')).toBeVisible();
  });
  it('when moves are set as an active tab, move cards should be visible and match the snapshot', () => {
    const { container } = render(customCardsRender(false, true, AvailableTabs.moves));
    const moveCards = screen.getAllByTestId('move-card-mock');
    movesMock.forEach(({ name }) => expect(screen.getByText(name)).toBeVisible());
    moveCards.forEach((card) => expect(card).toBeVisible());
    expect(moveCards).toHaveLength(movesMock.length);
    expect(container).toMatchSnapshot();
  });
  it('when types are set as an active tab, types cards should be visible and match the snapshot', () => {
    const { container } = render(customCardsRender(false, true, AvailableTabs.types));
    const typeCards = screen.getAllByTestId('type-card-mock');
    typesMock.forEach(({ name }) => expect(screen.getByText(name)).toBeVisible());
    typeCards.forEach((card) => expect(card).toBeVisible());
    expect(typeCards).toHaveLength(typesMock.length);
    expect(container).toMatchSnapshot();
  });
  it('when pokemons are set as an active tab, pokemons cards should be visible and match the snapshot', () => {
    const { container } = render(customCardsRender());
    const pokemonCards = screen.getAllByTestId('pokemon-card-mock');
    const pokemonImages = screen.getAllByTestId('pokemon-img-mock');
    pokemonsMock.forEach(({ name }) => expect(screen.getByText(name)).toBeVisible());
    pokemonImages.forEach((image) => expect(image).toBeVisible());
    pokemonCards.forEach((card) => expect(card).toBeVisible());
    expect(pokemonCards).toHaveLength(pokemonsMock.length);
    expect(container).toMatchSnapshot();
  });
});
