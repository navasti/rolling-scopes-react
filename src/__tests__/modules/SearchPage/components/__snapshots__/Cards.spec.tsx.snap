// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards when moves are set as an active tab, move cards should be visible and match the snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ MMRei"
    min="130px"
  >
    <div />
    <button
      data-testid="move-card-mock"
    >
      <p>
        water-gun
      </p>
    </button>
    <button
      data-testid="move-card-mock"
    >
      <p>
        disable
      </p>
    </button>
    <button
      data-testid="move-card-mock"
    >
      <p>
        tackle
      </p>
    </button>
  </div>
</div>
`;

exports[`Cards when pokemons are set as an active tab, pokemons cards should be visible and match the snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ dkoYgU"
    min="190px"
  >
    <div />
    <button
      data-testid="pokemon-card-mock"
    >
      <div>
        <img
          alt="bulbasaur"
          data-testid="pokemon-img-mock"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png"
        />
        <div>
          <span>
            Base:
          </span>
          <p>
            Name: 
            <span>
              bulbasaur
            </span>
          </p>
          <p>
            Types: 
            <span>
              fire
            </span>
          </p>
        </div>
      </div>
    </button>
    <button
      data-testid="pokemon-card-mock"
    >
      <div>
        <img
          alt="pikachu"
          data-testid="pokemon-img-mock"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png"
        />
        <div>
          <span>
            Base:
          </span>
          <p>
            Name: 
            <span>
              pikachu
            </span>
          </p>
          <p>
            Types: 
            <span>
              fire
            </span>
          </p>
        </div>
      </div>
    </button>
    <button
      data-testid="pokemon-card-mock"
    >
      <div>
        <img
          alt="charmander"
          data-testid="pokemon-img-mock"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png"
        />
        <div>
          <span>
            Base:
          </span>
          <p>
            Name: 
            <span>
              charmander
            </span>
          </p>
          <p>
            Types: 
            <span>
              fire
            </span>
          </p>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`Cards when types are set as an active tab, types cards should be visible and match the snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ MMRei"
    min="130px"
  >
    <div />
    <button
      data-testid="type-card-mock"
    >
      <p>
        fighting
      </p>
    </button>
    <button
      data-testid="type-card-mock"
    >
      <p>
        ground
      </p>
    </button>
    <button
      data-testid="type-card-mock"
    >
      <p>
        flying
      </p>
    </button>
  </div>
</div>
`;
